### PR TITLE
Move things to `rustc_type_ir`

### PR DESCRIPTION
**Pull request by compiler-errors**
Finishes some work proposed in https://github.com/rust-lang/compiler-team/issues/341.

r? @ghost

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -92,6 +91,70 @@ pub trait OnDiskCache<'tcx>: rustc_data_structures::sync::Sync {
     fn serialize(&self, tcx: TyCtxt<'tcx>, encoder: &mut FileEncoder) -> FileEncodeResult;
 }
 
+pub struct TyInterner<'tcx> {
+    pub tcx: TyCtxt<'tcx>,
+}
+
+/// We don't ever actually need this. It's only required for derives.
+impl<'tcx> Hash for TyInterner<'tcx> {
+    fn hash<H: Hasher>(&self, _state: &mut H) {}
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/context.rs:100]: Maybe panic in all these impls to make sure we don't actually use them

File: compiler/rustc_middle/src/ty/mod.rs [hunk 2]
```diff
@@ -463,16 +463,31 @@ pub(crate) struct TyS<'tcx> {
...
-    outer_exclusive_binder: DebruijnIndex::from_usize(0),
-};
+const LEAKED_BOOL_TY_ALREADY: std::sync::atomic::AtomicBool =
+    std::sync::atomic::AtomicBool::new(false);
+
+/// "Static" bool only used for internal testing.
+///
+/// FIXME(rustc_type_ir): This really should be replaced with something that doesn't leak.
+/// however, since it's used for testing, it's not _that_ bad.
+pub fn leak_bool_ty_for_unit_testing<'tcx>() -> Ty<'tcx> {
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/mod.rs:473]: Two alternatives I can think of (not mutually exclusive):
1) Move this closer to the use site.
2) If the test is under `cfg(test)`, then put this under `cfg(test)` too.

File: compiler/rustc_type_ir/src/sty.rs [hunk 3]
```diff
@@ -186,6 +186,42 @@ pub enum TyKind<I: Interner> {
     Error(I::DelaySpanBugEmitted),
 }
 
+#[allow(rustc::usage_of_ty_tykind)]
+impl<I: Interner> Clone for TyKind<I> {
```
Comment (jackh726) [inline compiler/rustc_type_ir/src/sty.rs:190]: This is because `I: Clone` doesn't hold, isn't it?

File: compiler/rustc_middle/src/ty/vtable.rs [hunk 4]
```diff
@@ -57,7 +58,7 @@ pub(super) fn vtable_allocation_provider<'tcx>(
 
         tcx.vtable_entries(trait_ref)
     } else {
-        COMMON_VTABLE_ENTRIES
+        common_vtable_entries()
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/vtable.rs:61]: I wonder if you could do `const COMMON_VTABLE_ENTRIES: &'tcx [VtblEntry<'tcx>] = common_vtable_entries(); COMMON_VTABLE_ENTRIES` here to guarantee that it gets const evaluated. (But perf will tell us if we need to try to be tricky).

File: compiler/rustc_middle/src/ty/mod.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/mod.rs:473]: I can't put this closer to the use site, because constructing an `Interned` requires a private constructor that's only accessible from middle, and the usage is in a different crate.

I can't put it under `cfg(test)` for the same reason -- `cfg(test)` doesn't propagate, right?

File: compiler/rustc_type_ir/src/sty.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/sty.rs:190]: Yup

File: compiler/rustc_middle/src/ty/vtable.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/vtable.rs:61]: I don't think consts can reference arbitrary inner lifetimes, can they?

Comment (jackh726) [reply compiler/rustc_middle/src/ty/vtable.rs:61]: Uh...probably not actually.

File: compiler/rustc_middle/src/ty/sty.rs [hunk 5]
```diff
@@ -78,6 +80,7 @@ impl BoundRegionKind {
     }
 }
 
+/*
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/sty.rs:83]: Can you remove this commented code now?

File: compiler/rustc_middle/src/ty/mod.rs [hunk 6]
```diff
@@ -447,30 +449,45 @@ pub(crate) struct TyS<'tcx> {
...
-static_assert_size!(TyS<'_>, 40);
+//#[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
+//static_assert_size!(TyS<'_>, 40);
 
 // We are actually storing a stable hash cache next to the type, so let's
 // also check the full size
-#[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
-static_assert_size!(WithStableHash<TyS<'_>>, 56);
+//#[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
+//static_assert_size!(WithStableHash<TyS<'_>>, 56);
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/mod.rs:458]: Any way we can add this back?

File: compiler/rustc_type_ir/src/sty.rs [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+use rustc_serialize::{Decodable, Encodable};
+
+/// Defines the kinds of types used by the type system.
+///
+/// Types written by the user start out as [hir::TyKind](rustc_hir::TyKind) and get
+/// converted to this representation using `AstConv::ast_ty_to_ty`.
+#[allow(rustc::usage_of_ty_tykind)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+//#[derive(TyEncodable, TyDecodable)]
+//#[derive(HashStable)]
```
Comment (jackh726) [inline compiler/rustc_type_ir/src/sty.rs:18]: Can you remove these and instead leave a comment on where to find these impls and why?

**Comment by compiler-errors** [comment]
@rustbot author

Will address @jackh726's comments and need to fix the clippy lint.

**Comment by jackh726** [comment]
I think the "Make Lazy not care about lifetimes until decode" commit can also be split out into a separate PR?

File: compiler/rustc_middle/src/ty/parameterized.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    type Value<'tcx> = crate::mir::Body<'tcx>;
+}
+
+impl ParameterizedOverTcx for crate::thir::abstract_const::Node<'static> {
+    type Value<'tcx> = crate::thir::abstract_const::Node<'tcx>;
+}
+
+impl ParameterizedOverTcx for crate::middle::exported_symbols::ExportedSymbol<'static> {
+    type Value<'tcx> = crate::middle::exported_symbols::ExportedSymbol<'tcx>;
+}
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/parameterized.rs:127]: I imagine that this could be folded in a macro?

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/parameterized.rs:127]: So I tried doing something like

`$($ident:ident ::)* $name:ident < $lt:lifetime >` and I was getting some macro rules parse ambiguity because it's ambiguously a path.

Path segments also include their lifetimes, so I can't just do `$path:path < $lt:lifetime >` either...

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/parameterized.rs:127]: I could possibly use some other sigil to separate the path segments, but that's kinda ugly so I didn't really give it a try.

Comment (jackh726) [reply compiler/rustc_middle/src/ty/parameterized.rs:127]: I wouldn't expect the macro to need to take anything expect for the name of the type?

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/parameterized.rs:127]: I made a macro version that takes just a single identifier in #97376, which suffices I guess.

File: compiler/rustc_middle/src/ty/vtable.rs (hunk 4 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/ty/vtable.rs:61]: Associated consts can, and inherent impl blocks can have associated consts

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/vtable.rs:61]: neat, maybe I'll associate these consts with TyCtxt then

File: compiler/rustc_middle/src/ty/context.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/context.rs:101]: I was not a fan of having these impls at all actually, so I just ended up removing these in favor of including (an edited version) of the derive macro expansions for the common traits on `TyKind`.

**Comment by compiler-errors** [comment]
@rust-timer queue

**Comment by jackh726** [comment]
I don't necessarily mind reviewing this. But given that I've written a non-trivial portion of this code, it might be best to have a second reviewer.

**Comment by compiler-errors** [comment]
I don't think this changes much but I added some inlines, so
@bors try @rust-timer queue

also idk if we want to roll for a reviewer or just have Jack assigned, so for now
r? @jackh726

File: compiler/rustc_borrowck/src/lib.rs [hunk 9]
```diff
@@ -98,7 +98,11 @@ struct Upvar<'tcx> {
     by_ref: bool,
 }
 
-const DEREF_PROJECTION: &[PlaceElem<'_>; 1] = &[ProjectionElem::Deref];
+/// Associate some local constants with the `'tcx` lifetime
+struct TyCtxtConsts<'tcx>(TyCtxt<'tcx>);
+impl<'tcx> TyCtxtConsts<'tcx> {
+    const DEREF_PROJECTION: &'tcx [PlaceElem<'tcx>; 1] = &[ProjectionElem::Deref];
+}
```
Comment (jackh726) [inline compiler/rustc_borrowck/src/lib.rs:105]: In theory, this could be moved to a separate PR, but I don't feel very strongly about it.

File: compiler/rustc_lint/src/internal.rs [hunk 10]
```diff
@@ -123,55 +123,115 @@ declare_lint_pass!(TyTyKind => [
...
-            }
+    fn check_path(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        path: &'tcx rustc_hir::Path<'tcx>,
+        _: rustc_hir::HirId,
+    ) {
+        if let Some(segment) = path.segments.iter().nth_back(1)
+        && let Some(res) = &segment.res
+        && lint_ty_kind_usage(cx, res)
```
Comment (jackh726) [inline compiler/rustc_lint/src/internal.rs:134]: Drive by: While you're here, you could convert this to let else?

File: compiler/rustc_middle/src/ty/context.rs [hunk 11]
```diff
@@ -92,6 +91,35 @@ pub trait OnDiskCache<'tcx>: rustc_data_structures::sync::Sync {
     fn serialize(&self, tcx: TyCtxt<'tcx>, encoder: &mut FileEncoder) -> FileEncodeResult;
 }
 
+pub struct TyInterner<'tcx> {
+    pub tcx: TyCtxt<'tcx>,
+}
+
+#[allow(rustc::usage_of_ty_tykind)]
+impl<'tcx> Interner for TyInterner<'tcx> {
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/context.rs:99]: The other thing we could do is just `impl Interner for TyCtxt`

File: compiler/rustc_query_system/src/ich/mod.rs [hunk 12]
```diff
@@ -17,3 +18,124 @@ pub const IGNORED_ATTRIBUTES: &[Symbol] = &[
     sym::rustc_partition_codegened,
     sym::rustc_expected_cgu_reuse,
 ];
+
+#[allow(rustc::usage_of_ty_tykind)]
+impl<'__ctx, I: rustc_type_ir::Interner> HashStable<StableHashingContext<'__ctx>>
```
Comment (jackh726) [inline compiler/rustc_query_system/src/ich/mod.rs:23]: I think this, along with the other "should be derives but we can't" impls, deserves to have a comment of why we're essentially replicating `derive`.

File: compiler/rustc_type_ir/src/lib.rs [hunk 13]
```diff
@@ -7,9 +8,145 @@ extern crate rustc_macros;
 
 use rustc_data_structures::stable_hasher::{HashStable, StableHasher};
 use rustc_data_structures::unify::{EqUnifyValue, UnifyKey};
+//use rustc_serialize::{Decodable, Encodable};
```
Comment (jackh726) [inline compiler/rustc_type_ir/src/lib.rs:11]: This can be removed.

File: compiler/rustc_type_ir/src/lib.rs [hunk 14]
```diff
@@ -7,9 +8,145 @@ extern crate rustc_macros;
...
+    type BinderListTy: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type ListTy: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type ProjectionTy: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type ParamTy: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type BoundTy: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type PlaceholderType: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type InferTy: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type DelaySpanBugEmitted: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
+    type PredicateKind: Clone + Debug + Hash + PartialEq + Eq;
+    type AllocId: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord;
```
Comment (jackh726) [inline compiler/rustc_type_ir/src/lib.rs:45]: Trait aliases to the rescue? (Don't change anything, just a thought)

File: compiler/rustc_type_ir/src/sty.rs [hunk 15]
```diff
@@ -0,0 +1,907 @@
...
+}
+
+impl<I: Interner> TyKind<I> {
+    #[inline]
+    pub fn is_primitive(&self) -> bool {
+        matches!(self, Bool | Char | Int(_) | Uint(_) | Float(_))
+    }
+}
+
+fn discriminant<I: Interner>(value: &TyKind<I>) -> usize {
```
Comment (jackh726) [inline compiler/rustc_type_ir/src/sty.rs:201]: Is there not an intristic for this?

File: compiler/rustc_type_ir/src/sty.rs [hunk 16]
```diff
@@ -0,0 +1,907 @@
...
+        Opaque(_, _) => 21,
+        Param(_) => 22,
+        Bound(_, _) => 23,
+        Placeholder(_) => 24,
+        Infer(_) => 25,
+        Error(_) => 26,
+    }
+}
+
+impl<I: Interner> Clone for TyKind<I> {
```
Comment (jackh726) [inline compiler/rustc_type_ir/src/sty.rs:233]: A comment explaining why this impl (and the ones below) exists would be helpful

**Review by jackh726 — CHANGES_REQUESTED** [review]
Just a few nits, but overall ready I think.

File: compiler/rustc_query_system/src/ich/mod.rs (hunk 12 above)
Comment (compiler-errors) [reply compiler/rustc_query_system/src/ich/mod.rs:23]: Ah yes, can do.

File: compiler/rustc_type_ir/src/sty.rs (hunk 15 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/sty.rs:201]: There is. I was hesitant to use `core::intrinsics::discriminant_value`, and `std::mem::discriminant` returns a type that is not `PartialOrd` which we need in the `PartialOrd` impl.

Comment (jackh726) [reply compiler/rustc_type_ir/src/sty.rs:201]: Okay, then please at least add a comment to that regard

File: compiler/rustc_type_ir/src/lib.rs (hunk 14 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/lib.rs:45]:
```
trait InternedValue: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord {}
impl<T: Clone + Debug + Hash + PartialEq + Eq + PartialOrd + Ord> InternedValue for T {}
```

would make it so that all of these are implied, right?

**Comment by compiler-errors** [comment]
Regarding the perf hit, it only shows up in one test `bitmaps`. I think it's ok to take that perf hit and deal with it later, unless someone (cc: @rust-lang/wg-compiler-performance, perhaps) knows how to dig deeper into this.

File: compiler/rustc_type_ir/src/lib.rs (hunk 14 above)
Comment (jackh726) [reply compiler/rustc_type_ir/src/lib.rs:45]: Yes, maybe. Like I said, don't worry about it here. Would rather have the very-clear duplication than some question of what's going on (particularly since there are ongoing effects to make things not `PartialOrd`/`Ord`

File: compiler/rustc_typeck/src/check/op.rs [hunk 17]
```diff
@@ -677,6 +677,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         operand_ty: Ty<'tcx>,
         op: hir::UnOp,
     ) -> Ty<'tcx> {
+        use rustc_type_ir::sty::TyKind::*;
```
Comment (fbstj) [inline compiler/rustc_typeck/src/check/op.rs:680]: is this redundant with line 20?

File: compiler/rustc_lint/src/internal.rs (hunk 10 above)
Comment (compiler-errors) [reply compiler/rustc_lint/src/internal.rs:134]: Er, I think it's clear as a let chain. Since we're still checking a boolean condition, we would have to still either have to keep the `struct_span_lint` in an `if` or do `if !lint_ty_kind_usage { return; }`, making this method a longer to save an indentation level?

File: compiler/rustc_middle/src/ty/mod.rs [hunk 18]
```diff
@@ -463,16 +463,31 @@ static_assert_size!(WithStableHash<TyS<'_>>, 56);
...
+/// FIXME(rustc_type_ir): This really should be replaced with something that doesn't leak.
+/// however, since it's used for testing, it's not _that_ bad.
+pub fn leak_bool_ty_for_unit_testing<'tcx>() -> Ty<'tcx> {
+    use std::sync::atomic::*;
+
+    if LEAKED_BOOL_TY_ALREADY.load(Ordering::Acquire) {
+        panic!("Can only leak one bool type, since its equality depends on its address");
+    } else {
+        LEAKED_BOOL_TY_ALREADY.store(true, Ordering::Release);
+    }
```
Comment (SkiFire13) [inline compiler/rustc_middle/src/ty/mod.rs:480]: Couldn't two threads both load `false` and then store `true`, thus creating two different values?

```suggestion
    if LEAKED_BOOL_TY_ALREADY.swap(true, Ordering::AcqRel) {
        panic!("Can only leak one bool type, since its equality depends on its address");
    }
```

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/mod.rs:480]: I'm probably going to get rid of this in favor of another solution, actually.

File: compiler/rustc_middle/src/ty/context.rs [hunk 19]
```diff
@@ -91,6 +90,31 @@ pub trait OnDiskCache<'tcx>: rustc_data_structures::sync::Sync {
...
+impl<'tcx> Interner for TyCtxt<'tcx> {
+    type AdtDef = ty::AdtDef<'tcx>;
+    type SubstsRef = ty::SubstsRef<'tcx>;
+    type DefId = DefId;
+    type Ty = Ty<'tcx>;
+    type Const = ty::Const<'tcx>;
+    type Region = Region<'tcx>;
+    type TypeAndMut = TypeAndMut<'tcx>;
+    type Mutability = hir::Mutability;
+    type Movability = hir::Movability;
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/context.rs:103]: We should probably check if we can start moving some of these types into rustc_type_ir

Comment (jackh726) [reply compiler/rustc_middle/src/ty/context.rs:103]: Yes, we can. When I originally wrote this, the idea was to be as small as possible by trading extra associated types.

**Comment by jackh726** [comment]
Yay
